src/core/lib/viewport-transition-interpolator.js

```diff
@@ -0,0 +1,185 @@
+  ['longitude', 'latitude', 'zoom', 'bearing', 'pitch'];
+
+/** Util functions */
+function lerp(start, end, step) {
+  if (Array.isArray(start)) {
+    return start.map((element, index) => {
+      return lerp(element, end[index], step);
+    });
+  }
+  return step * end + (1 - step) * start;
+}
+function zoomToScale(zoom) {
+  return Math.pow(2, zoom);
+}
+
+function scaleToZoom(scale) {
+  return Math.log2(scale);
+}
+
+export class ViewportTransitionInterpolator {
+
+  /**
+   * Checks if two sets of props need transition in between
+   * @param currentProps {object} - a list of viewport props
+   * @param nextProps {object} - a list of viewport props
+   * @returns {bool} - true if two props are equivalent
+   */
+  arePropsEqual(currentProps, nextProps) {
+    for (const key of VIEWPORT_TRANSITION_PROPS) {
+      if (!equals(currentProps[key], nextProps[key])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns viewport props in transition
+   * @param startProps {object} - a list of starting viewport props
+   * @param endProps {object} - a list of target viewport props
+   * @param t {number} - a time factor between [0, 1]
+   * @returns {object} - a list of interpolated viewport props
+   */
+  interpolateProps(startProps, endProps, t) {
+    // Dummy return value -- this should be overriden by child classes
+    return endProps;
+  }
+
+}
+
+/**
+ * Performs linear interpolation of two viewports.
+*/
+export class ViewportLinearInterpolator extends ViewportTransitionInterpolator {
+
+  /**
+   * @param {Array} transitionProps - list of props to apply linear transition to.
+   */
+  constructor(transitionProps = VIEWPORT_TRANSITION_PROPS) {
+    super();
+    this.transitionProps = transitionProps;
+  }
+
+  arePropsEqual(currentProps, nextProps) {
+    for (const key of this.transitionProps) {
+      if (!equals(currentProps[key], nextProps[key])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  interpolateProps(startProps, endProps, t) {
+    const viewport = {};
+
+    for (const key of this.transitionProps) {
+      const startValue = startProps[key];
+      let endValue = endProps[key];
+      assert(isValid(startValue));
+      assert(isValid(endValue));
+
+      if (isAngularProp(key) && Math.abs(endValue - startValue) > 180) {
+        endValue = (endValue < 0) ? endValue + 360 : endValue - 360;
+      }
+
+      viewport[key] = lerp(startValue, endValue, t);
+    }
+    return viewport;
+  }
+
+}
+
+/**
+ * This class adapts mapbox-gl-js Map#flyTo animation so it can be used in
+ * react/redux architecture.
+ * mapbox-gl-js flyTo : https://www.mapbox.com/mapbox-gl-js/api/#map#flyto.
+ * It implements “Smooth and efficient zooming and panning.” algorithm by
+ * "Jarke J. van Wijk and Wim A.A. Nuij"
+*/
+export class ViewportFlyToInterpolator extends ViewportTransitionInterpolator {
+
+  /* eslint-disable max-statements */
+  interpolateProps(startProps, endProps, t) {
```

Comment:
This is identical to the old `viewportFlyToInterpolator` function

Reply:
As far as I can tell there is not a single reference to `this` inside this method. To emphasize that it is stateless and can be tested functionally, I'd keep this big function on the global scope and just call it here.